libraries/custom_appbase/include/eosio/chain/exec_pri_queue.hpp

```diff
@@ -61,22 +76,56 @@ class exec_pri_queue : public boost::asio::execution_context
       should_exit_ = [](){ assert(false); return true; }; // should not be called when locking is disabled
    }
 
-   // called from appbase::application_base::exec poll_one() or run_one()
    template <typename Function>
-   void add(int priority, exec_queue q, size_t order, Function function) {
+   void add(int priority, exec_queue q, size_t order, Function&& function) {
       assert( num_read_threads_ > 0 || q != exec_queue::read_exclusive);
       prio_queue& que = priority_que(q);
-      std::unique_ptr<queued_handler_base> handler(new queued_handler<Function>(priority, order, std::move(function)));
+      std::unique_ptr<queued_handler_base> handler(new queued_handler<Function>(handler_id::unique, priority, order, std::forward<Function>(function)));
       if (lock_enabled_ || q == exec_queue::read_exclusive) { // called directly from any thread for read_exclusive
          std::lock_guard g( mtx_ );
-         que.push( std::move( handler ) );
+         que.push( handler.release() );
          if (num_waiting_)
             cond_.notify_one();
       } else {
-         que.push( std::move( handler ) );
+         que.push( handler.release() );
       }
    }
 
+   // called from appbase::application_base::exec poll_one() or run_one()
+   template <typename Function>
+   void add(handler_id id, int priority, exec_queue q, size_t order, Function&& function) {
+      assert( num_read_threads_ > 0 || q != exec_queue::read_exclusive);
+      if (id == handler_id::unique) {
+         return add(priority, q, order, std::forward<Function>(function));
+      }
+      prio_queue& que = priority_que(q);
+      std::unique_lock g( mtx_, std::defer_lock );
+      if (lock_enabled_ || q == exec_queue::read_exclusive) {
+         // called directly from any thread for read_exclusive
+         g.lock();
+      }
+      if (!que.empty()) {
+         // find the associated priority
+         auto end = que.ordered_end();
+         auto i = std::lower_bound(que.ordered_begin(), end, priority, [](const auto& h, int priority) {
+            return h->priority() > priority;
+         });
+         if (i != end) {
+            // ordered iterator appears to only be a forward iterator
+            auto p = i;
+            for (; i != end; p = i, ++i) {
+               if ((*i)->priority() != priority)
+                  break;
+            }
+            if ((*p)->priority() == priority && (*p)->id() == id)
+               return;
+         }
```

Comment:
Could it be as simple as this?

```suggestion
         if (i != end) {
            // ordered iterator appears to only be a forward iterator
            for (; i != end && (*i)->priority() == priority; ++i)
               if ((*i)->id() == id)
                  return;
         }
```

Reply:
No. You have to find the last posted handler with the same priority and see if that one has the same id.

Reply:
Why does it has to be the last posted one?

Reply:
This goes with the more conservative approach of only de-dup on matching id for a given priority when it was the last id posted.

Say you have items in the queue, u=unique, pb=process_incoming_block
`q9(u) q8(pb) q7(u)`

Say you want to post a `q6(pb)`, it is possible this is not safe to de-dup because if `q7(u)` does something that needs `q6(pb)` to run then you end up with `q6(pb)` not running. So this goes with the more conservative route of only de-dup if `q7(u)` is a `pb`. Granted we could argue that `q7(u)` should post a `pb` if needs that to happen.

Reply:
The code I sent dedups only if it is the same id (pb)  at the same priority, so in your example posting `q6(pb)` would not dedup as there is no other `q6(pb)` in the queue. Or am I missing something?

Reply:
The 9,8,7 are just order numbers. They are all the same priority.
`q9(u) q8(pb) q7(u)`
So your code would find `q8(pb)` and return (de-dup). Want to only compare to `q7(u)` and then non-dedup.

Reply:
Is there a reason for this requirement, i.e would the code not work if we didn't check that it was the last posted? It seems to me that either we want to dedup, and then we should just do it, or not.

Reply:
Been thinking about this.  Maybe I am being too conservative. If the last item running needs to have a follow non-unique item it should post it.